src/portable.rs

```diff
@@ -70,6 +72,157 @@ impl PortableRegistry {
     pub fn types(&self) -> &[PortableType] {
         &self.types
     }
+
+    /// Retains only the portable types needed to express the provided ids.
+    ///
+    /// Returns the type IDs that have been retained.
+    /// The order of the type IDs in the returned vector corresponds to their
+    /// new positions in the type registry after filtering.
+    ///
+    /// For instance, if the function returns the vector [30, 10, 20], it means that
+    /// type ID 30 is now at position 0 and has the type ID 0 in the registry,
+    /// type ID 10 is now type ID 1, and type ID 20 is now type ID 2.
+    ///
+    /// # Note
+    ///
+    /// A given type ID can be defined by nesting type IDs, such as the case
+    /// of a [`TypeDef::Composite`] and others. To retain a valid [`PortableRegistry`]
+    /// all the types needed to express an ID are included. Therefore, the number of
+    /// elements defined by the result equals or exceeds the number of provided IDs.
+    pub fn retain(
+        &mut self,
+        ids: impl IntoIterator<Item = u32>,
+    ) -> Result<Vec<u32>, PortableRegistryError> {
```

Comment:
Thinking about what `Vec::retain` etc looks like and the general usage, I was wondering whether a signature more like this would also work:

```rust
fn retain<F, R>(
    &mut self,
    filter: F,
    on_retained: R
) 
where
    F: FnMut(u32) -> bool,
    R: FnMut(u32, u32)
{ /*...*/ }
```

Where `F` is called for each type we're not sure about keeping (some types will be kept by virtue of being needed recursively by other types, after all) to ask the user whether to keep it or not. `R` is called for any type that's kept (including recursive ones) to note the old and new ID.

The implementation could then do one pass over the "old" registry, pushing types (and recursively pushing subtypes that are needed) to it as it goes.

This would remove the need for any return type (and no more PortableRegistryError, since the user doesn't get to choose what type IDs to look for; only which to filter). It'd also allow the user to collect the "old -> new" Id mappings via the `R` function however they please (ie push to a vec as you have done, or to a map or do some update in place or whatever).

But that all said, in reality I can see that the current approach would work just fine for what we need too (ie, we'd collect all types mentioned in the frame-metadata types and pass them into here, obtaining a mapping we can then use to go back over said frame-metadata stuff to do the updating). Just sortof pondering how I'd approach it I guess :)

Reply:
That makes total sense and would boost the flexibility of the API. That's implemented now :D Thanks!